### PR TITLE
[wip] Metadata task definition fields

### PR DESCRIPTION
**Pull request by jtoberon**
### Summary
Expose two more metadata fields, task definition family and task definition

### Implementation details
Retrieve the fields from the existing task struct, and copy them into the 
metadata JSON. 

### Testing

- [ ] Builds on Linux (`make release`)
- [ ] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [ ] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [ ] Integration tests on Linux (`make run-integ-tests`) pass
- [ ] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [ ] Functional tests on Linux (`make run-functional-tests`) pass
- [ ] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: yes

### Description for the changelog
Enhancement -Expose two more metadata fields, task definition family and task definition

### Licensing

This contribution is under the terms of the Apache 2.0 License: yes

### CODE REVIEW

File: agent/containermetadata/parse_metadata.go [hunk 1]
```diff
@@ -40,17 +42,26 @@ func (manager *metadataManager) parseMetadataAtContainerCreate(taskARN string, c
...
 		cluster:                 manager.cluster,
-		taskMetadata:            TaskMetadata{containerName: containerName, taskARN: taskARN},
+		taskMetadata:            TaskMetadata{containerName: containerName, taskARN: task.Arn, taskDefinitionFamily: task.Family, taskDefinitionRevision: taskDefinitionRevision},
 		dockerContainerMetadata: dockerMD,
 		containerInstanceARN:    manager.containerInstanceARN,
 		metadataStatus:          MetadataReady,
 	}
 }
 
+func parseTaskDefinitionRevision(task *api.Task, containerName string) int {
```
Comment (jhaynes) [inline agent/containermetadata/parse_metadata.go]: This needs a comment describing what it does. It is also probably worth mentioning that when `strconv.Atoi` fails with an error, `taskDefinitionRevision` will be set to the `int` `0`.

File: agent/containermetadata/types.go [hunk 2]
```diff
@@ -151,6 +155,8 @@ func (m Metadata) MarshalJSON() ([]byte, error) {
 			Cluster:              m.cluster,
 			ContainerInstanceARN: m.containerInstanceARN,
 			TaskARN:              m.taskMetadata.taskARN,
+			TaskDefinitionFamily: m.taskMetadata.taskDefinitionFamily,
+			TaskDefinitionRevision: m.taskMetadata.taskDefinitionRevision,
```
Comment (jhaynes) [inline agent/containermetadata/types.go]: nit: the spacing for all the other elements in the map should match this line.

File: agent/containermetadata/parse_metadata.go [hunk 3]
```diff
@@ -27,10 +28,11 @@ import (
 // available prior to container creation
 // Since we accept incomplete metadata fields, we should not return
 // errors here and handle them at this or the above stage.
-func (manager *metadataManager) parseMetadataAtContainerCreate(taskARN string, containerName string) Metadata {
+func (manager *metadataManager) parseMetadataAtContainerCreate(task *api.Task, containerName string) Metadata {
+	taskDefinitionRevision := parseTaskDefinitionRevision(task, containerName)
 	return Metadata{
 		cluster:              manager.cluster,
-		taskMetadata:         TaskMetadata{containerName: containerName, taskARN: taskARN},
+		taskMetadata:         TaskMetadata{containerName: containerName, taskARN: task.Arn, taskDefinitionFamily: task.Family, taskDefinitionRevision: taskDefinitionRevision},
```
Comment (aaithal) [inline agent/containermetadata/parse_metadata.go]: minor: can you split these into multiple lines?

File: agent/containermetadata/parse_metadata.go [hunk 4]
```diff
@@ -40,17 +42,31 @@ func (manager *metadataManager) parseMetadataAtContainerCreate(taskARN string, c
...
 // errors here and handle them at this or the above stage.
-func (manager *metadataManager) parseMetadata(dockerContainer *docker.Container, taskARN string, containerName string) Metadata {
-	dockerMD := parseDockerContainerMetadata(taskARN, containerName, dockerContainer)
+func (manager *metadataManager) parseMetadata(dockerContainer *docker.Container, task *api.Task, containerName string) Metadata {
+	taskDefinitionRevision := parseTaskDefinitionRevision(task, containerName)
+	dockerMD := parseDockerContainerMetadata(task.Arn, containerName, dockerContainer)
 	return Metadata{
 		cluster:                 manager.cluster,
-		taskMetadata:            TaskMetadata{containerName: containerName, taskARN: taskARN},
+		taskMetadata:            TaskMetadata{containerName: containerName, taskARN: task.Arn, taskDefinitionFamily: task.Family, taskDefinitionRevision: taskDefinitionRevision},
```
Comment (aaithal) [inline agent/containermetadata/parse_metadata.go]: same here. please split these into multiple lines (helps with readability)

File: agent/containermetadata/types.go [hunk 5]
```diff
@@ -130,34 +132,38 @@ type Metadata struct {
...
-	ImageID              string            `json:"ImageID,omitempty"`
-	ImageName            string            `json:"ImageName,omitempty"`
-	Ports                []api.PortBinding `json:"PortMappings,omitempty"`
-	Networks             []Network         `json:"Networks,omitempty"`
-	MetadataFileStatus   MetadataStatus    `json:"MetadataFileStatus,omitempty"`
+	Cluster                string            `json:"Cluster,omitempty"`
+	ContainerInstanceARN   string            `json:"ContainerInstanceARN,omitempty"`
+	TaskARN                string            `json:"TaskARN,omitempty"`
+	TaskDefinitionFamily   string            `json:"TaskDefinitionFamily,omitempty"`
+	TaskDefinitionRevision int             `json:"TaskDefinitionRevision,omitempty"`
```
Comment (aaithal) [inline agent/containermetadata/types.go]: I think this has to be `*int`. `0` is still a valid int and `omitempty` would not apply to it.

File: agent/containermetadata/parse_metadata.go [hunk 6]
```diff
@@ -40,17 +42,31 @@ func (manager *metadataManager) parseMetadataAtContainerCreate(taskARN string, c
...
 		metadataStatus:          MetadataReady,
 	}
 }
 
+// parseTaskDefinitionRevision converts the task.Version from a string to an int.
+// this allows us to match the type used in our CLI:
+// https://docs.aws.amazon.com/cli/latest/reference/ecs/describe-task-definition.html
+// if anything bad happens, then 0 is returned, so you must omitempty to omit this from
+// the metadata JSON output.
+func parseTaskDefinitionRevision(task *api.Task, containerName string) int {
```
Comment (aaithal) [inline agent/containermetadata/parse_metadata.go]: as per my comment in `containermetadata/types.go`, i think this has to return (*int, error). If there's an error, the pointer would not be set and omitempty will kick in.

File: agent/containermetadata/parse_metadata.go (hunk 3 above)
Comment (jtoberon) [reply agent/containermetadata/parse_metadata.go]: Yeah. I couldn't figure out what the line length convention for this repo is. I'll do something reasonable.

File: agent/containermetadata/mocks/containermetadata_mocks.go [hunk 7]
```diff
@@ -21,6 +21,7 @@ import (
 
 	go_dockerclient "github.com/fsouza/go-dockerclient"
 	gomock "github.com/golang/mock/gomock"
+	"github.com/aws/amazon-ecs-agent/agent/api"
```
Comment (ofiliz) [inline agent/containermetadata/mocks/containermetadata_mocks.go]: I think the convention is to order import statements as: 1) native go packages, 2) packages from the same repo, 3) other packages, with an empty line between each group. So this new line should be in 21?

**Comment by adnxn** [comment]
ping @jtoberon: what's going on with this PR? do we still need this merged?

**Comment by jtoberon** [comment]
I'll probably pick this back up in 2 weeks or so.

**Comment by adnxn** [comment]
closing in favor of https://github.com/aws/amazon-ecs-agent/pull/1295

**Pull request closed without merging**